nbodykit/extensionpoints.py

```diff
@@ -336,8 +337,180 @@ def open(self):
+        
+    @classmethod 
+    def print_help(cls):
+        return cls.parser.print_help()
+    
+algorithms = Namespace()
+
+class AlgorithmPluginMount(type):
+    """ Metaclass for an extension point that provides
+        the methods to manage
+        plugins attached to the extension point.
+    """
+    def __new__(cls, name, bases, attrs):
+        # for python 2, ensure extension points are objects
+        # this is important for python 3 compatibility.
+        if len(bases) == 0:
+            bases = (object,)
+        # Only add PluginInterface to the ExtensionPoint,
+        # such that Plugins will inherit from this.
+        if len(bases) == 1 and bases[0] is object:
+            bases = (AlgorithmPluginInterface,)
+        return type.__new__(cls, name, bases, attrs)
+
+    def __init__(cls, name, bases, attrs):
+
+        # only executes when processing the mount point itself.
+        if not hasattr(cls, 'plugins'):
+            cls.plugins = {}
+        # called for each plugin, which already has 'plugins' list
+        else:
+            if not hasattr(cls, 'plugin_name'):
+                raise RuntimeError("Plugin class must carry a plugin_name.")
+            
+            # register, if this plugin isn't yet
+            if cls.plugin_name not in cls.plugins:
+                # add a commandline argument parser that will parse
+                # the parameters needed to run the algorithm
+                cls.parser = PluginArgumentParser(cls.plugin_name, usage=None,
+                                            formatter_class=ArgumentDefaultsHelpFormatter)
+
+                # track names of classes
+                cls.plugins[cls.plugin_name] = cls
+                
+                # store as part of the algorithms namespace
+                setattr(algorithms, cls.plugin_name, cls)
+                
+                # try to call register class method
+                if hasattr(cls, 'register'): cls.register()
+                
+                # try to set the defaults of the argument parser from 
+                # the __init__ signature
+                cls._set_parser_defaults()
+
+    def create(kls, plugin_name, ns): 
+        """ 
+        Instantiate an `algorithm` plugin from 
+        this extension point, based on the Namespace
+        returned by a command-line parser
+
+        Parameters
+        ----------
+        ns: Namespace
+            the `argparse.Namespace` instance holding the 
+            arguments/keywords necessary to initialize
+        """
+        klass = kls.plugins[plugin_name]
         
-__all__ = ['DataSource', 'Painter', 'Transfer', 'MeasurementStorage']
+        args, kwargs = klass._namespace_to_args(ns)
+        return klass(*args, **kwargs)
+       
+                    
+def AlgorithmExtensionPoint(cls):
+    """ Declares a class as an extension point """
+    return add_metaclass(AlgorithmPluginMount)(cls)
+    
+@AlgorithmExtensionPoint
+class Algorithm:
```

Comment:
I've been thinking of a 'prepare' or 'initialize' stage for 'DataSource' (to read in BoxSize, set up units etc). Would something like this be useful for an algorithm as well?